### ISSUE TITLE
Handle case where server doesn't support pipelining?

### ISSUE THREAD

**Issue by ronag** [mcollina/undici#49]
Not sure exactly what happens when the server doesn't support pipelining and how to handle that in the client?

**Comment by mcollina** [comment]
How would the server communicate that back to the client?

**Comment by ronag** [comment]
> server may send a "Connection: Close" header after your first request is sent indicating it does not want to use a persistent connection.

**Comment by ronag** [comment]
Possibly an edge case not worth fixing... I think this mostly happens with very old servers.

**Comment by ronag** [comment]
So basically the logic here would be that the client does not pipeline until the first request has succeeded with a persistent connection in order to indicate that the server seems to support pipelining.

**Comment by mcollina** [comment]
That would be nice. However I fear it's a lot of complexity.
How are other clients doing it, if at all? Isn't this part of HTTP/1.1 that server has to support?

**Comment by ronag** [comment]
Probably, I’ll take a look. This might be another one of those edge cases.

Sorry to bring up every possible edge case. Just trying to be throughout.

**Comment by ronag** [comment]
> How are other clients doing it, if at all? Isn't this part of HTTP/1.1 that server has to support?

@szmarczak do you have any input?

**Comment by szmarczak** [comment]
> Isn't this part of HTTP/1.1 that server has to support?

No. Pipelining happens only on a persistent (keep-alive) connection and is **optional**. Unfortunately the HTTP/1.1 RFC does not specify how to verify whether the server supports pipelining or not and how many concurrent requests at maximum it should have.

**Comment by ronag** [comment]
> Unfortunately the HTTP/1.1 RFC does not specify how to verify whether the server supports pipelining or not and how many concurrent requests at maximum it should have.

I mean... you can kind of detect it:

> So basically the logic here would be that the client does not pipeline until the first request has succeeded with a persistent connection in order to indicate that the server seems to support pipelining.

**Comment by szmarczak** [comment]
> you can kind of detect it by sending a request to the server and see what it does.

One request is not sufficient. Because the connection is keep-alive it doesn't mean that it supports pipelining (according to the RFC). But I too think that most servers do support pipelining. To check if the server supports pipelining, you need to make two requests.

I see your point. I say let's assume that every `keep-alive` server does support pipelining, then only one request is enough to make the check.

**Comment by ronag** [comment]
I mean we could just explicitly state in the README that persistent connection and pipelining must be support. Is that good enough?

**Comment by szmarczak** [comment]
Yup. What happens otherwise? Does it throw when there already requests queued?

**Comment by ronag** [comment]
> Yup. What happens otherwise? Does it throw when there already requests queued?

Everything inflight fails except for the first request.

**Comment by szmarczak** [comment]
Seems to be fine :)
